dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java

```diff
@@ -0,0 +1,113 @@
+                PermissionNeeded config = permissionMap.computeIfAbsent(param._1,
+                        s -> { throw new IllegalStateException("No permission data for name: " + s); });
+                DvObject object = param._2;
+                result.add(createRestrictedObject(object, config));
+                if (config.needsOnOwner().length > 0) {
+                    result.add(createRestrictedOwnerObject(object, config));
+                }
+        };
+
+        return createNamedParameterStream(method, parameterValues)
+                .collect(HashSet::new, updateResultWithParameter, Set::addAll);
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private Map<String, PermissionNeeded> createPermissionMap(PermissionNeeded[] configuration) {
+        return Arrays.stream(configuration)
+                .collect(Collectors.toMap(PermissionNeeded::on, Function.identity()));
+    }
+
+    private PermissionNeeded[] extractAnnotations(Method method) {
+        return Option.of(method.getAnnotation(Restricted.class))
+                .map(Restricted::value)
+                .getOrElse(new PermissionNeeded[0]);
+    }
+
+    private RestrictedObject createRestrictedObject(DvObject object, PermissionNeeded config) {
+        return RestrictedObject.of(config.on(), object, toSet(config.needs()), config.allRequired());
+    }
+
+    private RestrictedObject createRestrictedOwnerObject(DvObject object, PermissionNeeded config) {
+        DvObject owner = object != null ? object.getOwner() : null;
+        return RestrictedObject.of(config.on() + "_owner", owner, toSet(config.needsOnOwner()), config.allRequired());
+    }
+
+    private Set<Permission> toSet(Permission[] array) {
+        return Arrays.stream(array)
+                .collect(Collectors.toSet());
+    }
+
+    private Stream<Tuple2<String, DvObject>> createNamedParameterStream(Method method, Object[] parameterValues) {
+        Annotation[][] annotations = method.getParameterAnnotations();
+        if (annotations.length < parameterValues.length) {
+            throw new IllegalStateException("Mismatch between annotation and parameter array size: too few parameters. " +
+                    "Cannot properly process annotation.");
+        }
+
+        return IntStream.range(0, parameterValues.length)
+                .mapToObj(i -> Tuple.of(annotations[i], parameterValues[i]))
+                .map(t -> Tuple.of(extractName(t._1), t._2))
+                .filter(t -> !Name.EMPTY.equals(t._1))
+                .map(t -> Tuple.of(t._1.name, (DvObject) t._2));
+    }
+
+    private Name extractName(Annotation[] annotations) {
+        return Arrays.stream(annotations)
+                .filter(a -> a instanceof PermissionNeeded)
+                .map(PermissionNeeded.class::cast)
+                .map(PermissionNeeded::value)
+                .map(Name::new)
+                .findFirst()
+                .orElse(Name.EMPTY);
```

Comment:
Maybe method & object name is a bit misleading in this case. The point here is to match parameter annotations (which are got by reflection on the method) with parameter values (which are provided by the interception context): in case there is no PermissionNeeded annotation on param, it's simply a parameter that is not a subject of access restriction.

Reply:
Ok. Thanks for clarification. I've read code the wrong way. `Name.EMPTY` strongly suggest that there is an annotation, but its value is empty.
I would consider either renaming `Name.Empty` to something that will suggest that it does not exists or make signature of method return `Optional<Name>` and not use constant at all